datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1507,6 +1515,9 @@ impl LogicalPlan {
                             (_, Some(dt)) => {
                                 param_types.insert(id.clone(), Some(dt.clone()));
                             }
+                            (Some(Some(_)), None) => {
+                                // we have already inferred the datatype
```

Comment:
When can this happen? Like what situations would one instance of the parameter like `$2` be resolved but another instance would not be 🤔

Reply:
This happens because we earlier treated LIMIT specially. We added it to the `param_types` but when we get to the Placeholder expression here, we see that the type was never inferred. If we weren't in an immutable method, we could do the actual type inference above instead of directly populating `param_types` and let it be appropriately populated here.